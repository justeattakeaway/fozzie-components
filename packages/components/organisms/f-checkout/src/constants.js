export const CHECKOUT_METHOD_COLLECTION = 'collection';
export const CHECKOUT_METHOD_DELIVERY = 'delivery';

export const VALID_CHECKOUT_METHODS = [CHECKOUT_METHOD_COLLECTION, CHECKOUT_METHOD_DELIVERY];

export const TENANT_MAP = {
    'en-GB': 'uk',
    'en-AU': 'au',
    'en-NZ': 'nz',
    'da-DK': 'dk',
    'es-ES': 'es',
    'en-IE': 'ie',
    'it-IT': 'it',
    'nb-NO': 'no'
};

export const VALIDATIONS = {
    address: 'addressValidations',
    guest: 'guestValidations'
};
