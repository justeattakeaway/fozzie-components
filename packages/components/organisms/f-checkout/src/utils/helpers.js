/**
 * Gets a specific cookie based on the cookie name you provide it.
 *
 * TODO: Use universal cookie helper when consuming apps are ready.
 *
 * @param name
 * @returns {*}
 */
const getCookie = name => {
    const nameEQ = `${name}=`;
    const ca = document.cookie.split(';');
    for (let i = 0; i < ca.length; i++) {
        let c = ca[i];
        while (c.charAt(0) === ' ') c = c.substring(1, c.length);
        if (c.indexOf(nameEQ) === 0) return c.substring(nameEQ.length, c.length);
    }
    return null;
};

export {
    // eslint-disable-next-line import/prefer-default-export
    getCookie
};
