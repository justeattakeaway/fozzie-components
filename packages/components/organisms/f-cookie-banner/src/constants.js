export const CONSENT_COOKIE_NAME = 'je-cookieConsent';
export const LEGACY_CONSENT_COOKIE_NAME = 'je-banner_cookie';
