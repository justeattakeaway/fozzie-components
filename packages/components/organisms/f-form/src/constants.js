export const FORM_EVENTS = {
    fieldUpdated: 'updated',
    submitting: 'form-submitting'
};

export const PROP_VALIDATION_MESSAGES = {
    requiredFormFields: '`formFields` Not found in `formData` prop, please add an array of `formFields`',
    invalidFormFields: '`formFields` are invalid, please add an array of formFields',
    requiredNameProperty: 'Each field in `formFields` should contain "name" property',
    requiredTranslationsProperty: 'Each field in `formFields` should contain "translations" property',
    requiredLabelProperty: 'Each field in `formFields` should contain "translations.label" property'
};

export const DEFAULT_BUTTON_TEXT = 'Submit';
