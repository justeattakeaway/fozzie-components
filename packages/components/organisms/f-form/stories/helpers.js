const firstNameData = {
    name: 'firstName',
    value: 'John',
    translations: {
        label: 'First Name',
        validationMessages: {
            required: 'Enter First Name'
        }
    }
};

const phoneNumberData = {
    name: 'mobileNumber',
    value: '07111111111',
    translations: {
        label: 'Mobile Number',
        validationMessages: {
            required: 'Enter Mobile Number',
            invalid: 'Enter valid Mobile Number'
        }
    }
};

const postcodeData = {
    name: 'postcode',
    value: 'EC1A1BB',
    translations: {
        label: 'Postcode',
        validationMessages: {
            required: 'Enter Postcode',
            invalid: 'Enter valid Postcode'
        }
    }
};

const formData = {
    formFields: [
        firstNameData,
        {
            name: 'lastName',
            value: 'Johnson',
            translations: {
                label: 'Last Name'
            }
        },
        {
            name: 'email',
            value: 'John.Johnson@gmail.com',
            translations: {
                label: 'Email Address',
                validationMessages: {
                    required: 'Enter Email Address',
                    invalid: 'Enter valid Email Address'
                }
            }
        },
        phoneNumberData,
        postcodeData
    ],
    buttonText: 'Continue'
};

export {
    formData,
    firstNameData,
    phoneNumberData,
    postcodeData
};
