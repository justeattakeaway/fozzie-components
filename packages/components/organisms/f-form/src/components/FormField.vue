<template>
    <form-field
        v-bind="fieldProps"
        @input="updateField({ fieldName: fieldData.name, value: $event })"
        @blur="hasInvalidErrorMessage && formFieldBlur()">
        <template
            v-if="hasError"
            #error>
            <error-message
                :id="translations.errorName"
                data-js-error-message
                :data-test-id="translations[`${errorType}Error`]">
                {{ translations.validationMessages[errorType] }}
            </error-message>
        </template>
    </form-field>
</template>

<script>
import FormField from '@justeat/f-form-field';
import ErrorMessage from '@justeat/f-error-message';
import { FORM_EVENTS } from '../constants';

export default {
    components: {
        ErrorMessage,
        FormField
    },

    props: {
        fieldData: {
            type: Object,
            required: true
        }
    },

    inject: ['$v'],

    computed: {
        kebabCase () {
            return this.fieldData.name.replace(/([a-z0-9]|(?=[A-Z]))([A-Z1-9])/g, '$1-$2').toLowerCase();
        },

        translations () {
            return {
                ...this.fieldData.translations,
                name: `${this.kebabCase}`,
                errorName: `${this.kebabCase}-error`,
                emptyError: `error-${this.kebabCase}-empty`,
                invalidError: `error-${this.kebabCase}-invalid`
            };
        },

        fieldProps () {
            let props = {
                name: this.translations.name,
                value: this.fieldData.value,
                'label-text': this.translations.label,
                'input-type': this.inputType,
                'max-length': this.fieldData.mexLength || null
            };

            if (this.ariaLabel) {
                props = {
                    ...props,
                    'aria-label': this.ariaLabel
                };
            }

            if (this.hasError) {
                props = {
                    ...props,
                    'has-error': true,
                    'aria-invalid': true,
                    'aria-describedby': this.translations.validationMessages[this.errorType]
                };
            }

            return props;
        },

        inputType () {
            const inputTypes = {
                mobileNumber: 'tel',
                email: 'email',
                default: 'text'
            };

            return inputTypes[this.fieldData.name] || inputTypes.default;
        },

        hasValidationMessages () {
            return !!this.translations?.validationMessages;
        },

        hasInvalidErrorMessage () {
            return !!this.translations?.validationMessages?.invalid;
        },

        hasError () {
            return !!this.hasValidationMessages && (this.hasRequiredError || this.hasInvalidError);
        },

        isEmpty () {
            return this.validations?.$dirty;
        },

        validations () {
            return this.$v.formFields[this.fieldData.name];
        },

        hasRequiredError () {
            return this.isEmpty && !this.validations.required;
        },

        hasInvalidError () {
            return !!this.hasInvalidErrorMessage && this.isEmpty && !this.validations[this.fieldData.name];
        },

        errorType () {
            if (this.hasError) {
                return this.hasRequiredError ? 'required' : 'invalid';
            }
            return null;
        },

        ariaLabel () {
            if (this.fieldData.name === 'mobileNumber' && this.fieldData.value) {
                return [...this.fieldData.value].join(' ') || null;
            }
            return null;
        }
    },

    methods: {
        updateField ({ fieldName, value }) {
            this.$emit(FORM_EVENTS.fieldUpdated, { fieldName, value });
        },

        formFieldBlur () {
            if (this.validations) {
                this.validations.$touch();
            }
        }
    }
};
</script>

<style lang="scss" module>
</style>
