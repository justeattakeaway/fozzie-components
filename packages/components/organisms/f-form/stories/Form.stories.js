import { withA11y } from '@storybook/addon-a11y';
import VForm from '../src/components/Form.vue';
import { formData } from './helpers';

export default {
    title: 'Components/Organisms',
    decorators: [withA11y]
};

export const VFormComponent = () => ({
    components: { VForm },

    data () {
        return {
            formData
        };
    },

    methods: {
        updateField ({ name, value }) {
            const formField = this.formData.formFields.find(field => field.name === name);
            formField.value = value;
        }
    },


    template: '<v-form :form-data="formData" @updated="updateField"/>'
});

VFormComponent.storyName = 'f-form';
