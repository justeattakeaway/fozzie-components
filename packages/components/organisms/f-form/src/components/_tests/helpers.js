import { createLocalVue } from '@vue/test-utils';
import { VueI18n } from '@justeat/f-globalisation';
import tenantConfigs from '../../tenants';
import { fieldWithRequiredValidations } from '../../../stories/helpers';

const localVue = createLocalVue();
localVue.use(VueI18n);

const i18n = {
    locale: 'en-GB',
    messages: {
        'en-GB': tenantConfigs['en-GB'].messages
    }
};

const $v = {
    formFields: {
        mobileNumber: {
            $dirty: false,
            mobileNumber: false
        },
        postCode: {
            $dirty: false,
            postCode: false
        },
        firstName: {
            $dirty: false,
            $touch: jest.fn()
        },
        lastName: {
            $dirty: false
        },
        email: {
            $dirty: false,
            email: false
        }
    },
    dirty: false,
    $touch: jest.fn()
};

const FormFieldMock = {
    props: {
        name: 'first-name',
        'label-text': 'First Name',
        'input-type': 'text'
    },
    translations: fieldWithRequiredValidations.translations,
    validationMessages: fieldWithRequiredValidations.translations.validationMessages,
    value: fieldWithRequiredValidations.value
};

const FormFieldErrorMock = {
    props: {
        maxLength: null,
        name: 'firstName',
        'label-text': 'First Name',
        'input-type': 'text',
        'max-length': null,
        'aria-invalid': true,
        'has-error': true,
        'aria-describedby': 'error-postcode-required'
    },
    errorMessage: {
        props: {
            'data-test-id': 'error-postcode-required',
            id: 'error-postcode-required'
        },
        text: fieldWithRequiredValidations.translations.validationMessages.required
    },
    translations: fieldWithRequiredValidations.translations,
    validationMessages: fieldWithRequiredValidations.translations.validationMessages,
    value: fieldWithRequiredValidations.value
};

export {
    $v,
    localVue,
    i18n,
    FormFieldMock,
    FormFieldErrorMock
};
