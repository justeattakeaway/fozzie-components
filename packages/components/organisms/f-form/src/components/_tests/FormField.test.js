import { shallowMount } from '@vue/test-utils';
import { $v } from './helpers';
import { firstNameData } from '../../../stories/helpers';
import FormField from '../FormField.vue';
import { FORM_EVENTS } from '../../constants';

describe('FormField', () => {
    const propsData = {
        fieldData: firstNameData
    };

    it('should be defined', () => {
        // Arrange
        const wrapper = shallowMount(FormField, {
            propsData,
            provide: () => ({
                $v
            })
        });

        // Assert
        expect(wrapper.exists()).toBe(true);
    });

    describe('computed ::', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });

        describe('kebabCase ::', () => {
            it.each([
                ['firstName', 'first-name'],
                ['line1', 'line-1'],
                ['city', 'city']
            ])('should convert `fieldName` value to kebab case', (fieldName, expected) => {
                // Arrange & Act
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        fieldData: {
                            name: fieldName
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.kebabCase).toEqual(expected);
            });
        });

        describe('translations ::', () => {
            it('should return translations for form field', () => {
                // Arrange & Act
                const wrapper = shallowMount(FormField, {
                    propsData,
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.translations).toMatchSnapshot();
            });
        });

        describe('fieldProps ::', () => {
            it('should return field props', () => {
                // Arrange & Act
                const wrapper = shallowMount(FormField, {
                    propsData,
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.fieldProps).toMatchSnapshot();
            });

            describe('when field has error ::', () => {
                it('should return field props with error props', () => {
                    // Arrange & Act
                    const wrapper = shallowMount(FormField, {
                        computed: {
                            hasError: () => true
                        },
                        propsData,
                        provide: () => ({
                            $v
                        })
                    });

                    // Assert
                    expect(wrapper.vm.fieldProps).toMatchSnapshot();
                });
            });
        });

        describe('inputType ::', () => {
            it.each([
                ['tel', 'mobileNumber'],
                ['email', 'email'],
                ['text', 'firstName']
            ])('should return %s when field name is %s', (expected, name) => {
                // Arrange & Act
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        fieldData: {
                            ...firstNameData,
                            name
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.inputType).toEqual(expected);
            });
        });

        describe('hasValidationMessages ::', () => {
            it('should return false if field does not have `validationMessages', () => {
                // Arrange & Act
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        ...propsData,
                        fieldData: {
                            ...firstNameData,
                            translations: {
                                label: 'First Name'
                            }
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.hasValidationMessages).toEqual(false);
            });

            it('should return true if field has `validationMessages', () => {
                // Arrange & Act
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        ...propsData,
                        fieldData: {
                            ...firstNameData,
                            translations: {
                                label: 'First Name',
                                validationMessages: {
                                    required: 'Enter First Name'
                                }
                            }
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.hasValidationMessages).toEqual(true);
            });
        });

        describe('hasInvalidErrorMessage ::', () => {
            it('should return false if field do not have an `invalid` `validationMessages', () => {
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        ...propsData,
                        fieldData: {
                            ...firstNameData,
                            translations: {
                                label: 'FirstName'
                            }
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.hasInvalidErrorMessage).toEqual(false);
            });

            it('should return false if field has an `invalid` `validationMessages', () => {
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        ...propsData,
                        fieldData: {
                            ...firstNameData,
                            translations: {
                                label: 'FirstName',
                                validationMessages: {
                                    invalid: 'Enter First Name'
                                }
                            }
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.hasInvalidErrorMessage).toEqual(true);
            });
        });

        describe('ariaLabel ::', () => {
            it.each([
                ['+ 4 4 7 1 1 1 1 1 1 1 1 1', '+447111111111'],
                [null, ''],
                [null, null]
            ])('should return %s if mobile vaule is %s', (expected, value) => {
                // Act
                const wrapper = shallowMount(FormField, {
                    propsData: {
                        ...propsData,
                        fieldData: {
                            ...propsData.fieldData,
                            value,
                            name: 'mobileNumber'
                        }
                    },
                    provide: () => ({
                        $v
                    })
                });

                // Assert
                expect(wrapper.vm.ariaLabel).toEqual(expected);
            });
        });
    });

    describe('methods ::', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });

        describe('updateField ::', () => {
            it('should emit `updated` event with fieldName and field value', () => {
                // Arrange
                const wrapper = shallowMount(FormField, {
                    propsData,
                    provide: () => ({
                        $v
                    })
                });

                const emitSpy = jest.spyOn(wrapper.vm, '$emit');
                const payload = { fieldName: 'firstName', value: 'Joe' };

                // Act
                wrapper.vm.updateField(payload);

                // Assert
                expect(emitSpy).toHaveBeenCalledWith(FORM_EVENTS.fieldUpdated, payload);
            });
        });
    });
});
