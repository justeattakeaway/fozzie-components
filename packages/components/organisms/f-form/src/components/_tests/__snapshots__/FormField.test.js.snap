// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField computed :: fieldProps :: should return field props 1`] = `
Object {
  "input-type": "text",
  "label-text": "First Name",
  "max-length": null,
  "name": "first-name",
  "value": "John",
}
`;

exports[`FormField computed :: fieldProps :: when field has error :: should return field props with error props 1`] = `
Object {
  "aria-describedby": undefined,
  "aria-invalid": true,
  "has-error": true,
  "input-type": "text",
  "label-text": "First Name",
  "max-length": null,
  "name": "first-name",
  "value": "John",
}
`;

exports[`FormField computed :: translations :: should return translations for form field 1`] = `
Object {
  "emptyError": "error-first-name-empty",
  "errorName": "first-name-error",
  "invalidError": "error-first-name-invalid",
  "label": "First Name",
  "name": "first-name",
  "validationMessages": Object {
    "required": "Enter First Name",
  },
}
`;
